src/clustering/kmeans/kmeans.cc

```diff
@@ -0,0 +1,462 @@
+                uint32_t* closest_centroids, size_t n_train, uint32_t random_state);
+
+    void
+    split_clusters(std::vector<int>& hassign, size_t n_train, uint32_t random_state);
+
+    void
+    elkan_L2(const DataType* x, const DataType* y, size_t d, size_t nx, size_t ny, uint32_t* ids, float* val);
+
+    void
+    exhaustive_L2sqr_blas(const DataType* x, const DataType* y, size_t d, size_t nx, size_t ny, uint32_t* ids,
+                          float* val);
+
+ private:
+    std::unique_ptr<DataType[]> centroids_ = nullptr;
+    std::shared_ptr<ThreadPool> pool_;
+    size_t num_clusters_;
+    size_t dim_;
+};
+
+}  // namespace knowhere
+
+namespace {
+#ifndef FINTEGER
+#define FINTEGER long
+#endif
+
+extern "C" {
+/* declare BLAS functions, see http://www.netlib.org/clapack/cblas/ */
+int
+sgemm_(const char* transa, const char* transb, FINTEGER* m, FINTEGER* n, FINTEGER* k, const float* alpha,
+       const float* a, FINTEGER* lda, const float* b, FINTEGER* ldb, float* beta, float* c, FINTEGER* ldc);
+
+#ifdef KNOWHERE_WITH_OPENBLAS
+void
+openblas_set_num_threads(int num_threads);
+#endif
+}
+}  // namespace
+
+namespace knowhere {
+
+template <typename DataType>
+expected<DataSetPtr>
+KmeansClusteringNode<DataType>::Train(const DataSet& dataset, const Config& cfg) {
+    auto kmeans_cfg = static_cast<const KmeansConfig&>(cfg);
+    auto num_clusters = kmeans_cfg.num_clusters.value();
+    auto random_state = kmeans_cfg.random_state.value();
+    auto max_iter = kmeans_cfg.max_iter.value();
+
+    auto rows = dataset.GetRows();
+    auto dim = dataset.GetDim();
+    auto vecs = dataset.GetTensor();
+    centroids_ = std::make_unique<DataType[]>(num_clusters * dim);
+    num_clusters_ = num_clusters;
+    dim_ = dim;
+    knowhere::TimeRecorder build_time("Kmeans train cost", 2);
+
+    initRandom((const DataType*)vecs, rows, random_state);
+    LOG_KNOWHERE_INFO_ << "num_clusters: " << num_clusters << " dim: " << dim;
+
+    float old_loss = std::numeric_limits<float>::max();
+    std::vector<std::vector<uint32_t>> closest_docs(num_clusters);
+    auto centroid_id_mapping = std::make_unique<uint32_t[]>(rows);
+
+    for (auto iter = 1; iter <= max_iter; ++iter) {
+        auto loss = lloyds_iter((const DataType*)vecs, closest_docs, centroid_id_mapping.get(), rows, random_state);
+
+        LOG_KNOWHERE_INFO_ << "Iter [" << iter << "/" << max_iter << "], loss: " << loss;
+        if ((loss < std::numeric_limits<float>::epsilon()) || ((iter != 1) && ((old_loss - loss) / loss) < 0)) {
+            LOG_KNOWHERE_INFO_ << "Residuals unchanged: " << old_loss << " becomes " << loss << ". Early termination.";
+            break;
+        }
+        old_loss = loss;
+    }
+    build_time.RecordSection("");
+    return GenResultDataSet(rows, centroid_id_mapping.release());
+}
+
+template <typename DataType>
+expected<DataSetPtr>
+KmeansClusteringNode<DataType>::Assign(const DataSet& dataset, const Config& cfg) {
```

Comment:
`cfg` is not used. Let's either mark it as `/* unused */` in the signature or get rid of this param